decisions/migrations.md

```diff
@@ -0,0 +1,262 @@
+  }
+
+  fn migrate(connection: &StorageConnection) -> anyhow::Error {
+    sql_query(&prepare).execute(&connection)?;
+    
+   let remote: Vec<RemoteBuffer> =  remote_buffer::dsl.load(connection);
+   let central: Vec<RemoteBuffer> =  remote_buffer::dsl.load(connection);
+   // Since buffer repository will now exist in code base, can use it from repository
+   BufferRepository::new(connection).upsert_many(remote.iter().map(/* translate to new Buffer */));
+   BufferRepository::new(connection).upsert_many(central.iter().map(/* translate to new Buffer */));
+
+    sql_query(&finalised).execute(&connection)?;
+  }
+}
+
+#[test]
+fn test() {
+   let con = init_database("test");
+   migrate(con, V1_01.version());
+
+   diesel::insert_into(remote_buffer::dsl::remote_buffer).values(&RemoteBuffer {
+    // Some mock values
+   }).execute(&con)?;
+
+   diesel::insert_into(central_buffer::dsl::central_buffer).values(&CentralBuffer {
+    // Some mock values
+   }).execute(&con)?;
+
+   migrate(con, V1_02.version());
+
+   assert_eq!(BufferRepository::new(connection).all(), vec![Buffer {/* some mock values */}, Buffer {/* some mock values */}];
+}
+```
+
+Above should explain how most of the requirements are met, `8.` can be met using cfg flags for pg/sqlite if they differ. If they differ only slightly like `item_stats/up.sql`, we can replace the difference (once again with cfg flags)
+
+It should be very easy to navigate to any given migration if they are located in their own file or directory. 
+
+`5.` can be dealt with by exporting to `schema spy` for each version, and hosting on github actions.
+`4.` As for logging, can be logged to files for now (but we do want further issue/discussion about logging) {insert issue here}
+
+*Pros:*
+- Ability to add tests (at the very least to avoid manual vaidation that data migrations work)
+- Flexibility and visibility of migration mechanism (can do more complex migrations, use transactions etc..)
+- Data manipulation consistency (as in we use diesel for most data manipulations now, can continue down that path for data migrations)
+
+*Cons:*
+- A bit more work upfront
+- Sligtly more overhead (although I think this is negligible)
+
+## Decision
+
+I suggest to go with `Option 2`, I don't think there is much work to make it happen (2-3 hours for the core and half a day to fill the gaps like logging etc..)
+
+## Other considerations
+
+TODO move this out to new issues / KDD or expand this KDD to include these
+
+### Integration test
+
+It would be great to have the ability to test migrations on live data file (annonymised). There is a challenge in validating that migrations are indeed correct, either with sanity check (which would require logic to be written and probably tested) or something like a data diff, the latter might be a good option to go with. If we use automated github actions for this it would be more `available` for tester and more likely to be looked at.
+
+### James comments
```

Comment:
I think most of these comments really relate to why having integration tests could be valuable  (e.g. they could identify performance issues with migrations, or might encounter database states that haven't been thought about in test cases) I think you could replace the 'James comments' section with something that :)

Reply:
Yeah, I've moved this to the epic, I think it's good to make a note of it somewhere but probably doesn't quite belong in KDD ?

I've removed and re-arranged some things: 9fb7df39277856529d63ed22a24bafc8e642230c
And added `Other` section in epic description: https://github.com/openmsupply/open-msupply/issues/704